postgres provider: allow to load layers with polyhedral surfaces or TINs

mhugo opened the pull request:
This restores a hack that was introduced prior to geometry refactoring in order to be able to load TIN and polyhedral surfaces layers in QGIS (in 2D).

@jef-n @mhugent just to be sure it does not break anything else ...

Would there be any way to setup a postgis provider unit tests ?

m-kuhn commented:
like this? https://github.com/qgis/QGIS/blob/master/tests/src/python/test_provider_postgres.py

strk commented:
is TRIANGLE already supported ? that's also an areal geometry

strk commented:
about that test: it might be useful for it to allow changing connection parameters via an env variable (libpq standard variables could do, beside database name) -- is that test already run on "make check" ?

mhugo commented:
@m-kuhn ah ! sorry I missed it :) I've just realised there is also a dedicated compilation option. Perfect. I'll have a look at it, thanks.

@strk you're right, TRIANGLE should also be handled. And I agree that the db parameters should be customizable

mhugo commented:
@m-kuhn Is the postgres provider test is run during a pull request build ? It should fail since it now relies on new tables ... which should not be there without your manual intervention. Right ?

m-kuhn commented:
@mhugo yes, they are run during pr builds. but since the db is recreated on every travis build, you can add your own extensions ( see https://github.com/qgis/QGIS/blob/master/ci/travis/linux/before_script.sh ) without my intervention.

mhugo commented:
@m-kuhn ah nice! So my new tests have successfully passed :)

mhugo commented:
@m-kuhn You added "Requires tests". Do you think this requires additional tests ? Or is it ok to merge before the freeze ?

m-kuhn commented:
Thanks for the tests.

Assigned @jef-n , he's most experienced with the postgres provider I think.

Commit:
PostGIS: Allow to load TIN, PS and Triangle layers

The postgres provider is modified so that layers with
TIN, PolyhedralSurface and Triangle geometries can be loaded.
Geometries are converted to MultiPolygons (and Polygons for Triangles).

The postgres test is completed to cover the loading of different types
of layers

mhugo commented:
Back on this topic.
This new version does not touch the geometry types, only the postgres provider. I think it's better, no new wkb type that are not really supported. It also fixes a part of the code in featureiterator that is now useless for most of the 2.5D geometries (only needed for TIN). And unit tests are now present.
Since it fixes a regression from 2.8, I'll merge it.
